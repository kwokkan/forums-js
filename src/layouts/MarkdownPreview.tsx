import { useState } from "react";
import { Input, Nav } from "rsuite";
import { Markdown } from "./Markdown";

interface IProps {
    content: string;
    name?: string;
    rows?: number;
    className?: string;
    placeholder?: string;
    onChange?: (content: string) => void;
}

const enum Mode {
    editor = "editor",
    preview = "preview",
}

function renderText(mode: Mode, props: IProps): React.ReactElement | null {
    if (mode !== Mode.editor) {
        return null;
    }

    function handleOnChange(value: string | number | ReadonlyArray<string>, _: React.SyntheticEvent): void {
        if (props.onChange) {
            props.onChange(value as string);
        }
    }

    return (
        <Input
            rows={props.rows}
            name={props.name}
            as="textarea"
            style={{ resize: "both" }}
            value={props.content}
            placeholder={props.placeholder}
            onChange={handleOnChange}
        />
    );
}

function renderPreview(mode: Mode, props: IProps): React.ReactElement | null {
    if (mode !== Mode.preview) {
        return null;
    }

    return (
        <div className="markdown-preview">
            <Markdown content={props.content} />
        </div>
    );
}

export function MarkdownPreview(props: IProps) {
    const [mode, setMode] = useState(Mode.editor);

    return (
        <div className={props.className}>
            <Nav activeKey={mode} appearance="subtle" onSelect={setMode}>
                <Nav.Item eventKey={Mode.editor} className="preview-strip-1">
                    Text
                </Nav.Item>
                <Nav.Item eventKey={Mode.preview} className="preview-strip-2">
                    Preview
                </Nav.Item>
            </Nav>

            {renderText(mode, props)}

            {renderPreview(mode, props)}
        </div>
    );
}
