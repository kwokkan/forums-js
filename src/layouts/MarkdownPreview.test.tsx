import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { act } from "react-dom/test-utils";
import { create } from "react-test-renderer";
import { MarkdownPreview } from "./MarkdownPreview";

beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
});

test("Renders without error - edit mode", () => {
    const content = "# header";
    const tree = create(
        <MarkdownPreview content={content} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders without error - preview mode", () => {
    const content = "# header";
    let wrapper = mount(
        <MarkdownPreview content={content} />
    );

    act(() => {
        wrapper.find("li.preview-strip-2 a").simulate("click");
    });

    wrapper = wrapper.update();

    expect(wrapper.debug()).toMatchSnapshot();
    expect(wrapper.find(".preview-strip-2 .rs-nav-item-active").length).toStrictEqual(1);
});
