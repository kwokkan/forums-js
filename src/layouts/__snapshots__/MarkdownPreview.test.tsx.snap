// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without error - edit mode 1`] = `
<div>
  <div
    className="rs-nav rs-nav-subtle rs-nav-horizontal"
  >
    <ul>
      <li
        className="rs-nav-item preview-strip-1 rs-nav-item-active"
      >
        <a
          className="rs-nav-item-content"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          Text
          <span
            className="rs-ripple-pond"
          >
            <span
              className="rs-ripple"
              style={Object {}}
            />
          </span>
        </a>
      </li>
      <li
        className="rs-nav-item preview-strip-2"
      >
        <a
          className="rs-nav-item-content"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          Preview
          <span
            className="rs-ripple-pond"
          >
            <span
              className="rs-ripple"
              style={Object {}}
            />
          </span>
        </a>
      </li>
    </ul>
    <div
      className="rs-nav-waterline"
    />
  </div>
  <textarea
    className="rs-input"
    id={null}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "resize": "both",
      }
    }
    type="text"
    value="# header"
  />
</div>
`;

exports[`Renders without error - preview mode 1`] = `
"<MarkdownPreview content=\\"# header\\">
  <div className={[undefined]}>
    <defaultProps(Nav) activeKey={2} appearance=\\"subtle\\" onSelect={[Function: bound dispatchAction]} classPrefix=\\"rs-nav\\">
      <Nav activeKey={2} appearance=\\"subtle\\" onSelect={[Function: bound dispatchAction]} classPrefix=\\"rs-nav\\">
        <div className=\\"rs-nav rs-nav-subtle rs-nav-horizontal\\">
          <ul>
            <defaultProps(NavItem) eventKey={1} className=\\"preview-strip-1\\" tabIndex={0} componentClass={{...}} classPrefix=\\"rs-nav-item\\" onSelect={[Function: bound dispatchAction]} hasTooltip={false} active={false}>
              <NavItem eventKey={1} className=\\"preview-strip-1\\" tabIndex={0} componentClass={{...}} classPrefix=\\"rs-nav-item\\" onSelect={[Function: bound dispatchAction]} hasTooltip={false} active={false}>
                <li className=\\"rs-nav-item preview-strip-1\\" style={[undefined]}>
                  <SafeAnchor disabled={[undefined]} role=\\"button\\" tabIndex={0} className=\\"rs-nav-item-content\\" onClick={[Function (anonymous)]}>
                    <a role=\\"button\\" tabIndex={0} className=\\"rs-nav-item-content\\" onClick={[Function: handleClick]}>
                      Text
                      <defaultProps(Ripple) classPrefix=\\"rs-ripple\\">
                        <Ripple classPrefix=\\"rs-ripple\\">
                          <span className=\\"rs-ripple-pond\\">
                            <Transition in={false} enteringClassName=\\"rs-ripple-rippling\\" onEntered={[Function (anonymous)]} timeout={1000}>
                              <span className=\\"rs-ripple\\" style={{...}} />
                            </Transition>
                          </span>
                        </Ripple>
                      </defaultProps(Ripple)>
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
            </defaultProps(NavItem)>
            <defaultProps(NavItem) eventKey={2} className=\\"preview-strip-2\\" tabIndex={0} componentClass={{...}} classPrefix=\\"rs-nav-item\\" onSelect={[Function: bound dispatchAction]} hasTooltip={false} active={true}>
              <NavItem eventKey={2} className=\\"preview-strip-2\\" tabIndex={0} componentClass={{...}} classPrefix=\\"rs-nav-item\\" onSelect={[Function: bound dispatchAction]} hasTooltip={false} active={true}>
                <li className=\\"rs-nav-item preview-strip-2 rs-nav-item-active\\" style={[undefined]}>
                  <SafeAnchor disabled={[undefined]} role=\\"button\\" tabIndex={0} className=\\"rs-nav-item-content\\" onClick={[Function (anonymous)]}>
                    <a role=\\"button\\" tabIndex={0} className=\\"rs-nav-item-content\\" onClick={[Function: handleClick]}>
                      Preview
                      <defaultProps(Ripple) classPrefix=\\"rs-ripple\\">
                        <Ripple classPrefix=\\"rs-ripple\\">
                          <span className=\\"rs-ripple-pond\\">
                            <Transition in={false} enteringClassName=\\"rs-ripple-rippling\\" onEntered={[Function (anonymous)]} timeout={1000}>
                              <span className=\\"rs-ripple\\" style={{...}} />
                            </Transition>
                          </span>
                        </Ripple>
                      </defaultProps(Ripple)>
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
            </defaultProps(NavItem)>
          </ul>
          <div className=\\"rs-nav-waterline\\" />
        </div>
      </Nav>
    </defaultProps(Nav)>
    <div className=\\"markdown-preview\\">
      <Markdown content=\\"# header\\">
        <ReactMarkdown source=\\"# header\\" renderers={{...}} escapeHtml={true} skipHtml={false} sourcePos={false} rawSourcePos={false} transformLinkUri={[Function: uriTransformer]} astPlugins={{...}} plugins={{...}} parserOptions={{...}}>
          <Root>
            <Heading level={1}>
              <h1>
                <TextRenderer nodeKey=\\"text-1-3-0\\" value=\\"header\\">
                  header
                </TextRenderer>
              </h1>
            </Heading>
          </Root>
        </ReactMarkdown>
      </Markdown>
    </div>
  </div>
</MarkdownPreview>"
`;
